File: cpp_10-99/30.cc
// C++
// leetcode 30
// https://leetcode-cn.com/problems/substring-with-concatenation-of-all-words/
// 滑动窗口

class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> res;
        int wordsNum = words.size();
        if(s.size() == 0 || wordsNum == 0)
            return  res;
        int wordSize = words[0].size();

        // 预处理
        map<string, int>    noSame;
        for(auto word : words)
            noSame[word]++;

        for(int i = 0; i < wordSize && i < s.size(); i++) {
            int left = i;
            int right = i;
            int cnt = 0;
            map<string, int>    cur;

            while(right + wordSize <= s.size()) {
                string curWord = s.substr(right, wordSize);
                right += wordSize;

                if(noSame.find(curWord) != noSame.end()) {
                    // 该词存在
                    cur[curWord]++;
                    cnt++;

                    while(cur[curWord] > noSame[curWord]) {
                        string toBeDelete = s.substr(left, wordSize);
                        left += wordSize;
                        cur[toBeDelete]--;
                        cnt--;
                    }

                    if(cnt == wordsNum)
                        res.push_back(left);
                } else {
                    // 该词不存在
                    left = right;
                    cnt = 0;
                    cur.clear();
                    if(left + wordsNum * wordSize > s.size())
                        continue;
                }
            }
        }

        return  res;
    }
};

File: Go/Go_10-99/30.go
// Go
// leetcode 30
// https://leetcode.cn/problems/substring-with-concatenation-of-all-words/

func findSubstring(s string, words []string) []int {
	n, wordNum, wordLength:= len(s), len(words), len(words[0])
	res := make([]int, 0, 5)

	// 
	need := make(map[string]int, wordNum)
	for _, word := range words {
		need[word]++
	}

	// working
	for i := 0; i < wordLength && i < n; i++ {
		l, r, cnt := i, i, 0
		cur := make(map[string]int)
		for r + wordLength <= n {
			curWord := s[r : r + wordLength]
			r += wordLength

			curNeed, ok := need[curWord]
			if ok {
				// need curWord
				cur[curWord]++
				cnt++

				for cur[curWord] > curNeed {
					toBeDelete := s[l : l + wordLength]
					l += wordLength
					cur[toBeDelete]--
					cnt--
				}

				if cnt == wordNum {
					res = append(res, l)
				}
			} else {
				// don't need curWord
				l = r
				cnt = 0
				cur = make(map[string]int)
				if l + wordNum * wordLength > n {
					continue
				}
			}
		}
	}
	return res
}